activitysim/abm/models/util/tour_od.py

```diff
@@ -0,0 +1,980 @@
+        chunk_size, trace_label):
+
+    chunk_tag = 'tour_od.sample'
+
+    origin_col_name = model_settings['ORIG_COL_NAME']
+    dest_col_name = model_settings['DEST_COL_NAME']
+    alt_dest_col_name = model_settings['ALT_DEST_COL_NAME']
+
+    skim_dict = network_los.get_default_skim_dict()
+    skims = skim_dict.wrap(origin_col_name, dest_col_name)
+
+    # the name of the od column to be returned in choices
+    alt_od_col_name = get_od_id_col(origin_col_name, dest_col_name)
+    choices = _od_sample(
+        spec_segment_name,
+        choosers,
+        network_los,
+        destination_size_terms,
+        origin_col_name,
+        dest_col_name,
+        skims,
+        estimator,
+        model_settings,
+        alt_od_col_name,
+        chunk_size,
+        chunk_tag,
+        trace_label)
+
+    choices[origin_col_name] = choices[alt_od_col_name].str.split('_').str[0].astype(int)
+    choices[dest_col_name] = choices[alt_od_col_name].str.split('_').str[1].astype(int)
+
+    return choices
+
+
+def map_maz_to_taz(s, network_los):
+    maz_to_taz = network_los.maz_taz_df[['MAZ', 'TAZ']].set_index('MAZ').TAZ
+    return s.map(maz_to_taz)
+
+
+def map_maz_to_ext_taz(s):
+    land_use = inject.get_table('land_use').to_frame(columns=['external_TAZ']).external_TAZ
+    return s.map(land_use).astype(int)
+
+
+def map_maz_to_ext_maz(s):
+    land_use = inject.get_table('land_use').to_frame(columns=['external_MAZ']).external_MAZ
+    return s.map(land_use).astype(int)
+
+
+def map_ext_maz_to_maz(s):
+    land_use = inject.get_table('land_use').to_frame(columns=['original_MAZ']).original_MAZ
+    return s.map(land_use).astype(int)
+
+
+def aggregate_size_terms(dest_size_terms, network_los):
+
+    # aggregate MAZ_size_terms to TAZ_size_terms
+    MAZ_size_terms = dest_size_terms.copy()
+
+    # add crosswalk DEST_TAZ column to MAZ_size_terms
+    MAZ_size_terms[DEST_TAZ] = map_maz_to_taz(MAZ_size_terms.index, network_los)
+
+    # aggregate to TAZ
+    TAZ_size_terms = MAZ_size_terms.groupby(DEST_TAZ).agg({'size_term': 'sum'})
+    assert not TAZ_size_terms['size_term'].isna().any()
+
+    #           size_term
+    # dest_TAZ
+    # 2              45.0
+    # 3              44.0
+    # 4              59.0
+
+    # add crosswalk DEST_TAZ column to MAZ_size_terms
+    # MAZ_size_terms = MAZ_size_terms.sort_values([DEST_TAZ, 'size_term'])  # maybe helpful for debugging
+    MAZ_size_terms = MAZ_size_terms[[DEST_TAZ, 'size_term']].reset_index(drop=False)
+    MAZ_size_terms = MAZ_size_terms.sort_values([DEST_TAZ, 'zone_id']).reset_index(drop=True)
+
+    #       zone_id  dest_TAZ  size_term
+    # 0        6097         2       10.0
+    # 1       16421         2       13.0
+    # 2       24251         3       14.0
+
+    # print(f"TAZ_size_terms ({TAZ_size_terms.shape})\n{TAZ_size_terms}")
+    # print(f"MAZ_size_terms ({MAZ_size_terms.shape})\n{MAZ_size_terms}")
+
+    return MAZ_size_terms, TAZ_size_terms
+
+
+def choose_MAZ_for_TAZ(
```

Comment:
Does this model only work with 2+ zone systems? I don't see any logic that would avert the need for MAZ and relying instead wholly on a one zone system.

Reply:
I would say roughly 70-80% of this code is shared with one of [location_choice.py](https://github.com/ActivitySim/activitysim/blob/master/activitysim/abm/models/location_choice.py#L316), [tour_destination.py](https://github.com/ActivitySim/activitysim/blob/master/activitysim/abm/models/util/tour_destination.py#L209), or [trip_destination.py](https://github.com/ActivitySim/activitysim/blob/master/activitysim/abm/models/trip_destination.py#L162). So the short answer is that I don't have any reason to believe that this model is any less-compatible with one zone systems than those. That being said, I can't say with 100% certainty that it would work without testing it against a proper one zone model, which we haven't done here.

Reply:
Was there anything in particular that stood out to you here as potentially more problematic for one zone systems than those other models I mentioned?

Reply:
Its the references to the maz system in general. If a MAZ system is not specified in the setup, what happens when the code attempts to developed a bridge between the TAZ and the seemingly null MAZ setup in a one zone system?

Reply:
OK, that's what I thought. I think your question is a good one and I can't say that I have a definitive answer to it as we have not explicitly tested for this scenario. However, if you look at where the references to MAZ's appear, they should be restricted entirely to the scenario in which choice sets are constructive via "pre-sampling" as opposed to the standard sampling mechanism. And its true that pre-sampling only works for multi-zone systems, as the main difference between pre-sampling and regular sampling is that with pre-sampling the choice set is first constructed by sampling alts by TAZ, and then randomly assigning a MAZ within each TAZ based on size terms. In short, the MAZ references should only be encountered if the user has turned on pre-sampling, which they should only do if they are using a multi-zone system. And if you look at the other models I linked to in my previous comment, that's exactly how they work as well.

That being said, there are two places in the tour OD code where I can find a reference to the MAZ system outside of the pre-sampling code. The first is [here](https://github.com/ActivitySim/activitysim/blob/fa898d6b1df524602760a010453d023892520459/activitysim/abm/models/util/tour_od.py#L161-L162), and the other is [here](https://github.com/ActivitySim/activitysim/blob/fa898d6b1df524602760a010453d023892520459/activitysim/abm/models/util/tour_od.py#L920-L926).

If I cleaned up those two MAZ-references, would that satisfy your concern about this model working for one-zone systems?